lib/ansible/utils/ssh_agent.py

```diff
@@ -0,0 +1,673 @@
+            digest.digest(),
+            newline=False
+        ).rstrip(b'=').decode('utf-8')
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class RSAPublicKeyMsg(PublicKeyMsg):
+    type: KeyAlgo
+    e: mpint
+    n: mpint
+    comments: unicode_string = dataclasses.field(default=unicode_string(''), compare=False)
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class DSAPublicKeyMsg(PublicKeyMsg):
+    type: KeyAlgo
+    p: mpint
+    q: mpint
+    g: mpint
+    y: mpint
+    comments: unicode_string = dataclasses.field(default=unicode_string(''), compare=False)
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class EcdsaPublicKeyMsg(PublicKeyMsg):
+    type: KeyAlgo
+    ecdsa_curve_name: unicode_string
+    Q: binary_string
+    comments: unicode_string = dataclasses.field(default=unicode_string(''), compare=False)
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class Ed25519PublicKeyMsg(PublicKeyMsg):
+    type: KeyAlgo
+    enc_a: binary_string
+    comments: unicode_string = dataclasses.field(default=unicode_string(''), compare=False)
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class KeyList(Msg):
+    nkeys: uint32
+    keys: PublicKeyMsgList
+
+    def __post_init__(self) -> None:
+        if self.nkeys != len(self.keys):
+            raise SshAgentFailure(
+                "agent: invalid number of keys received for identities list"
+            )
+
+
+@dataclasses.dataclass(order=True, slots=True)
+class PublicKeyMsgList(Msg):
+    keys: list[PublicKeyMsg]
+
+    def __iter__(self) -> t.Iterator[PublicKeyMsg]:
+        yield from self.keys
+
+    def __len__(self) -> int:
+        return len(self.keys)
+
+    @classmethod
+    def from_blob(cls, blob: memoryview | bytes) -> t.Self:
+        ...
+
+    @classmethod
+    def consume_from_blob(cls, blob: memoryview | bytes) -> tuple[t.Self, memoryview | bytes]:
+        args: list[PublicKeyMsg] = []
+        while blob:
+            prev_blob = blob
+            key_blob, key_blob_length, comment_blob = cls._consume_field(blob)
+
+            peek_key_algo, _length, _blob = cls._consume_field(key_blob)
+            pub_key_msg_cls = PublicKeyMsg.get_dataclass(
+                KeyAlgo(bytes(peek_key_algo).decode('utf-8'))
+            )
+
+            _fv, comment_blob_length, blob = cls._consume_field(comment_blob)
+            key_plus_comment = (
+                prev_blob[4: (4 + key_blob_length) + (4 + comment_blob_length)]
+            )
+
+            args.append(pub_key_msg_cls.from_blob(key_plus_comment))
+        return cls(args), b""
+
+    @staticmethod
+    def _consume_field(
+            blob: memoryview | bytes
+    ) -> tuple[memoryview | bytes, uint32, memoryview | bytes]:
+        length = uint32.from_blob(blob[:4])
+        blob = blob[4:]
+        return blob[:length], length, blob[length:]
+
+
+class SshAgentClient:
+    def __init__(self, auth_sock: str) -> None:
+        self._sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
```

Comment:
I see no timeouts on the `recv`s below- should we set something short and sane for the socket overall, or on the individual `recv`s- something buggy/broken seems like it could hang for a long time or maybe even forever...

Reply:
I set 10s timeout on the socket, let me know if different value would be more sane.